### ISSUE TITLE
on  Activation Checkpointing

### ISSUE THREAD

**Issue by ChaunceyWang** [microsoft/DeepSpeed#5704]
I  want to use [Activation Checkpointing](https://www.deepspeed.ai/docs/config-json/) in my project, as the activation values cost much GPU memory,I need to significantly reduce GPU memory usage.
When I set the config json like this, The memory remains the same.  It doesn't work!
```
  "activation_checkpointing": {
    "partition_activations": true,
    "cpu_checkpointing": true,
    "contiguous_memory_optimization": true,
    "number_checkpoints": 100,
    "synchronize_checkpoint_boundary": false,
    "profile": false
    }
```
Can someone give me a Example?  The [official tutorial](https://deepspeed.readthedocs.io/en/stable/activation-checkpointing.html) is very vague……，Thanks！

**Comment by ChaunceyWang** [comment]
Can you help me？ many people have questions about this，very confused. @adk9 @wjessup @dfyz @manuelciosici

**Comment by tjruwase** [comment]
@ChaunceyWang, activation checkpointing requires changes in the training code. The tutorial page pointed to required changes for (now deprecated) [Megatron-LM code](https://www.deepspeed.ai/tutorials/megatron/#deepspeed-activation-checkpoints-optional). If using HF trainer, you can enable the native support of this feature (called gradient checkpointing) described [here](https://huggingface.co/docs/transformers/v4.18.0/en/performance#gradient-checkpointing). If not, then consider studying the implementation in the Megatron-DeepSpeed code starting [here](https://github.com/microsoft/Megatron-DeepSpeed/blob/ea4b67a2bb5e572bdbe9c4fd3aa0e539e8e17110/megatron/arguments.py#L1499) and [here](https://github.com/microsoft/Megatron-DeepSpeed/blob/7eb36a11b3a9c48ed07b93692ccf22bfb5577f7e/megatron/model/transformer.py#L1966). That implementation is integrated to the DeepSpeed features highlighted in your original post.

**Comment by ChaunceyWang** [comment]
Thank you @tjruwase 
[Megatron-LM codes](https://www.deepspeed.ai/tutorials/megatron/#deepspeed-activation-checkpoints-optional)  are tedious and cumbersome.  HF trainer can't meet my requirements.  Why not write a concise tutorial/code based on deepspeed config?

my codes(on video generation, ~1.5B model parameters ) are based on torch and deepspeed, and enabled zero3 stage.  However,  intermediate activation values cost much gpu memory(~78G), as video resolution is 1920*1080

Any good suggestions?  How to place intermediate activation values on multiple GPUs or CPU?  That's why I need [Activation Checkpointing](https://www.deepspeed.ai/docs/config-json/) of deepspeed.

**Comment by tjruwase** [comment]
@ChaunceyWang, unfortunately activation checkpointing has to be manually implemented in your code first before it can take advantage of the DeepSpeed optimizations. In my experience the implementation tend to be module-specific, and thus difficult to generalize. This is why I referred you to Megatron-DeepSpeed which has such an implementation that is integrated with the DeepSpeed optimizations. Unfortunately, we have not had bandwidth to write a concise tutorial. 

On the other hand, It is possible that a simple (torch only) implementation like the one in this [bert example](https://github.com/microsoft/DeepSpeedExamples/blob/bbab278112dc56d68c0d203f501b53d860788375/training/bing_bert/nvidia/modelingpreln.py#L619) would be sufficient for your case. Either way, you will need to implement wrappers for your module forward logic.

**Comment by ChaunceyWang** [comment]
Thank you @tjruwase 
I know how to use[ torch.utils.checkpoint](https://pytorch.org/docs/stable/checkpoint.html) or [deepspeed.checkpointing.checkpoint](https://deepspeed.readthedocs.io/en/latest/activation-checkpointing.html) in my module, and that works .

However, GPU memory usage is still relatively large after using torch.utils.checkpoint. Are there examples that Implements **partition_activations** or **cpu_checkpointing**? Maybe they can reduce memory usage of one GPU, right?

**Comment by tjruwase** [comment]
@ChaunceyWang, since you are already using DS checkpointing, you are already half-way there. Now you need to enable the DS optimizations during training initialization. See this Megatron-DeepSpeed [example](https://github.com/microsoft/Megatron-DeepSpeed/blob/7eb36a11b3a9c48ed07b93692ccf22bfb5577f7e/megatron/initialize.py#L166) of how to enable the extra DS optimizations.

**Comment by tjruwase** [comment]
@ChaunceyWang, closing due to lack of response. Please re-open if needed.
